File: JavaPatterns/src/main/java/structural/Adapter.java
package structural;

import java.util.logging.Level;

/**
 * The adapter pattern converts interface of a class into expected interface,
 * allowing classes with incompatible interfaces to work together through inheritance.
 *
 * @author Daniel Leon
 */
//Adaptor
interface Log {
    void warning(String message);
    void error(String message);
}

//Adaptee
final class Logger {
    void log(Level level, String message) {
        System.out.printf("[%s] Logging message %s\n", level, message);
    }
}

//Adapter
class LoggerToLogAdapter implements Log {
    private final Logger logger;

    public LoggerToLogAdapter(Logger logger) { this.logger = logger; }

    public void warning(String message) {
        logger.log(Level.WARNING, message);
    }

    public void error(String message) {
        logger.log(Level.SEVERE, message);
    }
}

public class Adapter {

    public static void main(String[] args) {

        Log log = new LoggerToLogAdapter(new Logger());

        log.error("Error message in Java");
        log.warning("Warning message in java");
    }
}

File: ScalaPatterns/src/main/scala/structural/Adapter.scala
package structural

import java.util.logging.Level

/**
 * In Scala, we have a built-in concept of interface adapters, expressed as implicit classes.
 * When expected type of expression is Log, yet a Logger instance is used, Scala compiler will automatically wrap that instance in the adapter class.
 *
 * @author Daniel Leon
 */
object Adapter {

  trait Log {
    def warning(message: String)
    def error(message: String)
  }

  final class Logger {
    def log(level: Level, message: String) { printf("[%s] Logging message %s\n" , level, message) }
  }

  implicit class LoggerToLogAdapter(logger: Logger) extends Log {
    def warning(message: String) { logger.log(Level.WARNING, message) }
    def error(message: String) { logger.log(Level.SEVERE, message) }
  }

  def main(args:Array[String]) {
    val log: Log = new Logger()

    log.error("Error message in Scala")
    log.warning("Warning message in Scala")
  }
}
